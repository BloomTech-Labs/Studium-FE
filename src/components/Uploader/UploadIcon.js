import React from 'react';
import styled from 'styled-components';
import {ReactComponent as UploadButton} from '../../svgs/ImageUploadButton.svg';
import {useAppHooks} from '../../customHooks/useAppHooks.js';
/**
 * Upload Icon
 * @component
 * @example
 * return (
 *  <UploadIcon />
 * )
 */

export const UploadIcon = props => {
  const {appView} = useAppHooks();

  return (
    <StyledUploadIcon
      appView={appView}
      data-testid="upload-icon"
    ></StyledUploadIcon>
  );
};

const StyledUploadIcon = styled(UploadButton)`
  display: block;
  z-index: 25;
  position: relative;
  width: ${props => (props.appView === 'APP_VIEW_MOBILE' ? '67px' : '95px')};
  height: ${props => (props.appView === 'APP_VIEW_MOBILE' ? '59px' : '107px')};
  background-color: transparent;
`;

UploadIcon.propTypes = {};
