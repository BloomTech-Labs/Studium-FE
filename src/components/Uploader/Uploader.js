import React, {useEffect, useState} from 'react';
import styled from 'styled-components';
import {Upload, message} from 'antd';
import {uploadImage} from '../../actions/photo';
import {UploadIcon} from './UploadIcon.js';
import * as PropTypes from 'prop-types';
import {useAppHooks} from '../../customHooks/useAppHooks.js';

/**
 * Image Uploader
 * @component
 * @example
 * return (
 *  <Uploader id={"1"} />
 * )
 */
export const Uploader = props => {
  /**
   * @type PhotoReducerState
   */
  const {photosState, usersState, dispatch} = useAppHooks('Uploader');
  const [photoObject, setPhotoObject] = useState(false);

  useEffect(() => {
    Object.values(photosState.photos).forEach(photoObject => {
      if (photoObject.id === props.id) {
        setPhotoObject(photoObject);
      }
    });
  }, [photosState, props.id]);

  const customRequest = file => {
    file.id = props.id;
    dispatch(uploadImage(file));
  };

  const getUrl = () => {
    if (photoObject) {
      if (photoObject.file.uid) {
        return (
          'https://res.cloudinary.com/www-synapsapp-com/image/upload/v1582468332/' +
          photoObject.file.uid
        );
      }
    }
    return undefined;
  };

  const loadImage = photoObject && photoObject.file && photoObject.file.uid;
  return (
    <div data-testid="upload">
      <StyledUpload
        name="file"
        listType="picture-card"
        showUploadList={false}
        beforeUpload={beforeUpload}
        customRequest={customRequest}
      >
        {loadImage ? (
          <img
            src={getUrl()}
            alt="avatar"
            style={{width: '100%'}}
            data-testid="upload-image"
          />
        ) : (
          <UploadIcon />
        )}
      </StyledUpload>
    </div>
  );
};

Uploader.propTypes = {
  id: PropTypes.string,
};

const StyledUpload = styled(Upload)`
  && {
    > .ant-upload.ant-upload-select-picture-card {
      border: none;
      background-color: transparent;
      width: 67px;
      height: 59px;
    }
  }
`;

function beforeUpload(file) {
  const isJpgOrPng = file.type === 'image/jpeg' || file.type === 'image/png';
  if (!isJpgOrPng) {
    message.error('You can only upload JPG/PNG file!');
  }
  const isLt2M = file.size / 1024 / 1024 < 2;
  if (!isLt2M) {
    message.error('Image must smaller than 2MB!');
  }
  return isJpgOrPng && isLt2M;
}
