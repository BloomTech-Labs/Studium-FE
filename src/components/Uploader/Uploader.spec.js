import React from 'react';
import {
  customRender, getNodesByType, getByTestId, fireEvent, store, getByRole,
} from '../../utilities/test-utils.js';
import { Uploader } from './Uploader.js';
import moxios from 'moxios';

/**
 * Test the styled uploader
 */
describe( 'Styled Uploader', () => {
  //Test that styled uploaded hasn't changed sience last snapshot.
  test( 'snapshot renders', async() => {
    //Call custom render to wrap the component in fake providers.
    const { debug, container } = customRender( <Uploader id={ 1 }/> );
    
    // log out the component to the console when debug is turned on in env

    
    // get the button from the container.
    const button = await getByTestId( container, 'upload' );
    
    // test button against snapshot
    expect( button ).toMatchSnapshot();
  } );
  
  test( 'To have a click button.', async done => {
    // create fake file.
    const file = new File( [ '(⌐□_□)' ], 'chucknorris.png', {
      type: 'image/png',
    } );
    
    // install moxios to catch the axios request going out.
    moxios.install();
    moxios.withMock( () => {
      // render the component
      const { container, debug } = customRender( <Uploader id={ 1 }/> );
      logOutMessageOrDebug( { debug } );
      
      // get state from the store
      const { photosState } = store.getState();
      
      // get all the html elements needed to upload the file
      let button = getByRole( container, 'button' );
      let uploadIcon = getByTestId( container, 'upload-icon' );
      let inputNode = getNodesByType( container, 'input' )[ 0 ];
      
      // check to make sure state is empty
      expect( photosState.photos ).toEqual( {} );
      // add the file to the input
      fireEvent.change( inputNode, { target: { files: [ file ] } } );
      // fire the upload event
      fireEvent.click( button );
      
      moxios.wait( () => {
        try{
          // check the axios request
          let request = moxios.requests.mostRecent();
          // respond with fake message from server
          request
            .respondWith( {
              status: 201, response: {
                photo: { public_id: file.name, photo_url: file.name },
              },
            } )
            .then( () => {
              logOutMessageOrDebug( { debug } );
              logOutMessageOrDebug( { message: photosState.toString() } );
              // get the image from the dom
              const avatar = getByTestId( container, 'upload-image' );
              expect( avatar ).toBeInTheDocument();
              expect( uploadIcon ).not.toBeInTheDocument();
              expect( photosState.photos[ 1 ].file ).toEqual( {
                url: file.name, public_id: file.name,
              } );
              // call done for async events.
              done();
            } );
        }catch( e ){
          done.fail();
        }
      } );
    } );
  } );
  
  afterEach( () => {
    moxios.uninstall();
  } );
} );
