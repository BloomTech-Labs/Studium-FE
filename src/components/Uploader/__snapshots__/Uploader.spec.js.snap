// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled Uploader snapshot renders 1`] = `
.c1 {
  display: block;
  z-index: 25;
  position: relative;
  width: 67px;
  height: 59px;
  background-color: transparent;
}

.c0.c0 > .ant-upload.ant-upload-select-picture-card {
  border: none;
  background-color: transparent;
  width: 67px;
  height: 59px;
}

<div
  data-testid="upload"
>
  <span
    class="c0 ant-upload-picture-card-wrapper"
  >
    <div
      class="ant-upload ant-upload-select ant-upload-select-picture-card"
    >
      <span
        class="ant-upload"
        role="button"
        tabindex="0"
      >
        <input
          accept=""
          style="display: none;"
          type="file"
        />
        <svg
          class="c1"
          data-testid="upload-icon"
        >
          ImageUploadButton.svg
        </svg>
      </span>
    </div>
  </span>
</div>
`;
